### PR TITLE
Simplify tools/Makefile

### PR DESCRIPTION
**Pull request by stedolan**
I found the metaprogramming / escaping / eval in this file hard to follow, so here's a more straightforward implementation.

cc @dra27

### CODE REVIEW

File: tools/Makefile [hunk 1]
```diff
@@ -84,19 +46,39 @@ VPATH := $(filter-out -I,$(INCLUDES))
...
+programs_byte := \
+  ocamldep ocamlprof ocamlcp ocamloptp ocamlmklib  \
+  ocamlmktop ocamlcmt dumpobj ocamlobjinfo \
+  primreq stripdebug cmpbyt
+install_files += $(filter $(installed_tools), $(programs_byte))
+programs_opt := $(programs_byte:%=%.opt)
+
+programs_byte_exe := $(programs_byte:%=%$(EXE))
+programs_opt_exe := $(programs_opt:%=%$(EXE))
+$(programs_byte) $(programs_opt): %: $(if $(EXE), %$(EXE), )
```
Comment (dra27) [inline tools/Makefile]:
```suggestion
ifneq "$(EXE)" ""
$(programs_byte) $(programs_opt): %: %$(EXE)
endif
```

File: tools/Makefile [hunk 2]
```diff
@@ -138,9 +124,10 @@ installopt::
 	  "$(INSTALL_LIBDIR)"
 
 # To help building mixed-mode libraries (OCaml + C)
+ocamlmklib_cmos = config.cmo build_path_prefix_map.cmo misc.cmo ocamlmklib.cmo
```
Comment (dra27) [inline tools/Makefile]: `ocamlmklib_objects`?

File: tools/Makefile [hunk 3]
```diff
@@ -84,19 +46,39 @@ VPATH := $(filter-out -I,$(INCLUDES))
...
+$(programs_byte_exe):
+	$(CAMLC) $(LINKFLAGS) -I $(ROOTDIR) -o $@ $(filter-out %.cmi,$^)
+
+$(programs_opt_exe):
+	$(CAMLOPT_CMD) $(LINKFLAGS) -I $(ROOTDIR) -o $@ $(filter-out %.cmi,$^)
+
+all: $(programs_byte_exe)
+opt.opt: $(programs_opt_exe)
+clean::
+	rm -f $(programs_byte_exe) $(programs_opt_exe)
```
Comment (dra27) [inline tools/Makefile]: This isn't quite right - the `clean` target can't depend `$(EXE)`. `clean` should _always_ erase `ocamldep`, `ocamldep.opt`, `ocamldep.exe` and `ocamldep.opt.exe` (i.e. 4 files).

File: tools/Makefile [hunk 4]
```diff
@@ -84,19 +46,39 @@ VPATH := $(filter-out -I,$(INCLUDES))
...
   $(ROOTDIR)/compilerlibs/ocamlbytecomp.cma
-ocamldep$(EXE): LINKFLAGS += -compat-32
-$(call byte_and_opt,ocamldep,$(CAMLDEP_IMPORTS) $(CAMLDEP_OBJ),)
-ocamldep$(EXE): depend.cmi
-ocamldep.opt$(EXE): depend.cmi
 
-clean::
-	rm -f ocamldep ocamldep.exe ocamldep.opt ocamldep.opt.exe
+ocamldep$(EXE): LINKFLAGS += -compat-32
+ocamldep$(EXE): depend.cmi $(CAMLDEP_IMPORTS) $(CAMLDEP_OBJ)
```
Comment (dra27) [inline tools/Makefile]: You included a `filter-out` in the rule, so I think you can put `depend.cmi` in `CAMLDEP_IMPORTS` (or `ocamldep_objects`)

File: tools/Makefile [hunk 5]
```diff
@@ -109,16 +91,20 @@ CSLPROF_IMPORTS=config.cmo build_path_prefix_map.cmo misc.cmo identifiable.cmo \
   pprintast.cmo \
   lexer.cmo parse.cmo
 
-$(call byte_and_opt,ocamlprof,$(CSLPROF_IMPORTS) profiling.cmo $(CSLPROF),)
+ocamlprof$(EXE): $(CSLPROF_IMPORTS) profiling.cmo $(CSLPROF)
+ocamlprof.opt$(EXE): \
+  $(call byte2native, $(CSLPROF_IMPORTS) profiling.cmo $(CSLPROF))
```
Comment (dra27) [inline tools/Makefile]:
```suggestion
ocamlprof$(EXE): $(CSLPROF_IMPORTS) $(CSLPROF)
ocamlprof.opt$(EXE): \
  $(call byte2native, $(CSLPROF_IMPORTS) $(CSLPROF))

all: profiling.cmo
allopt: profiling.cmx
```
This has been wrong since #587 introduced it - `profiling.cmo` should always been in the third argument. It needs to built (and installed) for `ocamlcp` and `ocamloptp`, but not linked into `ocamlprof`

File: tools/Makefile [hunk 6]
```diff
@@ -84,19 +46,39 @@ VPATH := $(filter-out -I,$(INCLUDES))
 .PHONY: all allopt opt.opt # allopt and opt.opt are synonyms
 allopt: opt.opt
```
Comment (dra27) [inline tools/Makefile]: It's presently "wrong", but it would make sense _to move_ these two lines down below the `all:` and `opt.opt` lines ~below~ and make `all` the default target (at presently, confusingly, `allopt` is the default target)

**Review by dra27 — COMMENTED** [review]
Overall I think this is a good idea - there are a few things which definitely need tweaking before it's equivalent.

I would also make the variables more consistent - e.g. merge `CAMLDEP_OBJ`/`CAMLDEP_IMPORTS` and `CSLPROF`/`CSLPROF_IMPORTS`, given that each recipe has to repeat the objects. Seems sensible that `tool` has `tools_objects := ...`.

That said, the removal of macros could go further if you have `tool_imports` containing just `.cma` files and `tools_objects` containing just `.cmi` & `.cmo` files then `byte2native` could go, with `$(tool_imports:.cma=.cmxa) $(tool_objects:.cmo=.cmx)` in its place.

**Comment by gasche** [comment]
Ping: this PR should probably be brought to a conclusion before it bitrots away.

File: tools/Makefile (hunk 1 above)
Comment (stedolan) [reply tools/Makefile]: Hah, yes, that is clearer.

File: tools/Makefile (hunk 3 above)
Comment (stedolan) [reply tools/Makefile]: Fixed. (I had thought the old behaviour was a mistake)

File: tools/Makefile (hunk 6 above)
Comment (stedolan) [reply tools/Makefile]: I think you accidentally a verb?

File: tools/Makefile (hunk 5 above)
Comment (stedolan) [reply tools/Makefile]: Done

File: tools/Makefile (hunk 4 above)
Comment (stedolan) [reply tools/Makefile]: Done

File: tools/Makefile (hunk 2 above)
Comment (stedolan) [reply tools/Makefile]: > Overall I think this is a good idea - there are a few things which definitely need tweaking before it's equivalent.
> 
> I would also make the variables more consistent - e.g. merge `CAMLDEP_OBJ`/`CAMLDEP_IMPORTS` and `CSLPROF`/`CSLPROF_IMPORTS`, given that each recipe has to repeat the objects. Seems sensible that `tool` has `tools_objects := ...`.
> 
> That said, the removal of macros could go further if you have `tool_imports` containing just `.cma` files and `tools_objects` containing just `.cmi` & `.cmo` files then `byte2native` could go, with `$(tool_imports:.cma=.cmxa) $(tool_objects:.cmo=.cmx)` in its place.

I've removed the redundant variables but kept `byte2native` - I think it's clear and useful, it's a pure function with no weird eval / side-effects / quoting.

File: tools/Makefile (hunk 3 above)
Comment (gasche) [reply tools/Makefile]: The reason is that people may want to clean in a not-yet-configured or broken-configuration state. (Maybe we should have a comment to point that out.)

File: tools/Makefile (hunk 6 above)
Comment (dra27) [reply tools/Makefile]: Ah, "to move", etc.. I regularly from verb-disappearance and words order a funny in.

Comment (dra27) [reply tools/Makefile]: The key point was to make `all` the first target 🙂

File: tools/Makefile (hunk 2 above)
Comment (dra27) [reply tools/Makefile]: Agreed - up to you if you want to make the names themselves consistent between tools:

- s/stripdebug/STRIPDEBUG
- s/primreq/PRIMREQ
- s/OBJINFO/OCAMLOBJINFO
- s/ocamlmklib_cmos/OCAMLMKLIB
- s/ocamlcp_cmos/OCAMLCP
- s/CSLPROF/OCAMLPROF
- s/CAMLDEP/OCAMLDEP

**Comment by shindere** [comment]
@stedolan: what a briliant idea to simplify this file! Thank you so much!

Should you do other PRs that touch the build system, would you please mind
mentionning me so that I get a chance to review? I am not watching
the repo.

will start to review right now, many thinaks, @dra27, for having
given me the opportunity to do so!

**Comment by shindere** [comment]
Here are a few suggestions:

```
MAKEFLAGS := -r -R
```

How about removing this?

For the record, `-r` is for `--no-builtin-rules` and `-R` is for
`--no-builtin-variables`. If we really think these are important, I'd
say it should be stated e.g. in `Makefile.common`. Also defining the
variable explicitly that way feels wrong because it's make which is
supposed to deal with it. I also think the definition is wrong, becuase
according to make's manual, if, say, `-j -k` are given on make's
comand-line, the `MAKEFLAGS` will contain `jk`.

And this is already a sub-makefile, so it's not calling other makefiles
in turn, as far as I know,so i'd get rid of this. what do you think
@stedolan and @dra27?

**Comment by shindere** [comment]
How about also getting rid of the `DESTDIR ?=` line? It's defined the
same way in `Makefile.common` already...

**Comment by shindere** [comment]
```
override installed_tools := ocamldep ocamlprof ocamlcp ocamloptp \
                   ocamlmktop ocamlmklib ocamlobjinfo
```

How about getting rid of that "override" thing?

I must say I still don't find the variables easy to follow but it's not
to say the modification is not worth it, wuite the countrary. I think
the difficult to read bits were already there and Stephen's modification
does improve the situation.

I also think the `all` target should be defined first so that it becomes
the default one.

Regarding the definitions of `CAMLC` and `CAMMLOPT`, would itbe possible
to re-oder the options so that `-g` appears later? The idea behind this
request is to make it possible to use another compiler than the in-tree
one without necessary loosing the `-g`, if it is found important that it
stays here at all.

Regarding the way the `.exe` extension is dealt with, have you
considered using the `PROGRAM_SYNONYM` macro defined in
`Makefile.common`?

It's jsut a quesiton, just wondering whether you'd find the code more
readable with this macro?

Lines like

```
all: profiling.cmo
allopt: profiling.cmx
```

look wrong to me. When a target is defined in several places, shouldn't
the definitions use `::` rather than just `:` ?

**Comment by shindere** [comment]
Same question about the line

opt.opt: $(caml_tex)```
```

Shouldn't it rather be

```
opt.opt:: $(caml_tex)
```

with two `:` rather than one?

As a last question, but this is just out of curiosity and should
definitely not block Stephen's work to be merged: does someone know why
`lintapidiff` is built only as a nativ-code program? Beyond the program
itself, the concern behind my question is about regularity. Is there a
reason why this must be different from how other programs are compiled,
with rules for both a bytecode and a native version...

Thanks again for the work, @stedolan!

**Comment by dra27** [comment]
@shindere - the `all: profiling.cmo` and `allopt: profiling.cmx` entries don't have recipes, so they're just adding dependencies to the target. You only need the `::` (everywhere) if you want to append to the _recipe_.

In this case, I think it's better to have the fact that `profiling.cmo` is always built recorded with the rest of `ocamlprof`'s definitions, rather than at the top of the file (it's rather easier to use `grep` to search for all the dependencies of `all:`)

**Comment by shindere** [comment]
OK David, fine with me.

File: tools/Makefile (hunk 2 above)
Comment (stedolan) [reply tools/Makefile]: Done. I think the CAMLDEP -> OCAMLDEP change fixed a bug.

File: tools/Makefile (hunk 6 above)
Comment (stedolan) [reply tools/Makefile]: Done, explanation for the thanks.

**Comment by stedolan** [comment]
@shindere Thanks for the suggestions, I've applied them except for the one about `-g` (which I didn't really understand)

**Comment by stedolan** [comment]
@shindere re-reading your comment, I now understand what you meant about `-g`. I think I'd prefer not to make this change, though: someone using another compiler will already have to carefully reproduce the `-nostdlib`, `-I`, etc. options, and I don't think there's anything special about `-g`.

**Comment by shindere** [comment]
Stephen Dolan (2021/04/21 07:14 -0700):
> @shindere Thanks for the suggestions, I've applied them except for the
> one about `-g` (which I didn't really understand)

Thanks!

Sorry for the `-g`, let me try again and if it's still not clear, never
mind, let's leave it for a future PR.

One may want to compile OCaml with an olready installed compiler.
Something like

```
make OCAMLC=~/bin/ocamlc
```

(it's not quite possible yet but it's a desirable aim.)

When doing so, with the current definitioin, the `-g` flag would be
lost, whereas one may want to keep it (or at least, we don't want to
lose it just because we use another compiler than the default one).

Do I make more sense?

As I said, if not, not a big deal.

**Comment by shindere** [comment]
Stephen Dolan (2021/04/21 07:18 -0700):
> @shindere re-reading your comment, I now understand what you meant
> about `-g`. I think I'd prefer not to make this change, though:
> someone using another compiler will already have to carefully
> reproduce the `-nostdlib`, `-I`, etc. options, and I don't think
> there's anything special about `-g`.

Well, somehow the `-nostlid` business is indeed tightly boudn with which
compiler one is using. So IMO these chould appear early, in the compiler
definition itself, so that when changing the compiler all this gets
overriden. But if you don't want to make the change that's okay.

File: tools/Makefile [hunk 7]
```diff
@@ -83,44 +43,67 @@ COMPFLAGS = -absname -w +a-4-9-41-42-44-45-48-70 -strict-sequence \
...
   camlinternalMenhirLib.cmo parser.cmo \
   pprintast.cmo \
-  lexer.cmo parse.cmo
+  lexer.cmo parse.cmo ocamlprof.cmo
 
-$(call byte_and_opt,ocamlprof,$(CSLPROF_IMPORTS) profiling.cmo $(CSLPROF),)
+ocamlprof$(EXE): $(OCAMLPROF)
+ocamlprof.opt$(EXE): $(call byte2native, $(OCAMLPROF))
+all: profiling.cmo
+allopt: profiling.cmx
```
Comment (dra27) [inline tools/Makefile]:
```suggestion
opt.opt: profiling.cmx
```
I apologise - the top of the file says that `allopt` is an alias of `opt.opt` and my previous suggestion broke that...

File: tools/Makefile [hunk 8]
```diff
@@ -83,44 +43,67 @@ COMPFLAGS = -absname -w +a-4-9-41-42-44-45-48-70 -strict-sequence \
...
+          main_args.cmo
 
-$(call byte_and_opt,ocamlcp,$(ocamlcp_cmos) ocamlcp.cmo,)
-$(call byte_and_opt,ocamloptp,$(ocamlcp_cmos) ocamloptp.cmo,)
+ocamlcp$(EXE): $(OCAMLCP) ocamlcp.cmo
+ocamlcp.opt$(EXE): $(call byte2native, $(OCAMLCP) ocamlcp.cmo)
+ocamloptp$(EXE): $(OCAMLCP) ocamloptp.cmo
+ocamloptp.opt$(EXE): $(call byte2native, $(OCAMLCP) ocamloptp.cmo)
 
 opt:: profiling.cmx
```
Comment (dra27) [inline tools/Makefile:110]:
```suggestion
```
(not yours, but it's wrong)

**Review by dra27 — COMMENTED** [review]
Sorry, I was about to approve it and on a last read-through spotted a mistake, which was my suggestion... those are _definitely_ from me, I promise!

I agree about leaving the `-g` either for another day or completely - it's definitely not part of simplifying :o)

**Comment by gasche** [comment]
There is a small change proposal from @dra27 (which indeed fixes a small issue with the PR) but this is otherwise good to go.

**Comment by dra27** [comment]
My suggestion to remove the `opt:: profiling.cmx` was clearly nonsense!

There's something else which was already not right in that `ocamloptp` gets built even in bytecode-only mode (but without `profiling.cmx`) so there is some further tidying to do, but this PR should be got in quickly, before someone fiddles with the rest of the build system...

**Comment by dra27** [comment]
Good to squash when AppVeyor catches up

**Comment by dra27** [comment]
It's a build system PR, so I've given one last round of [precheck](https://ci.inria.fr/ocaml/job/precheck/594/) just to be sure.

**Comment by stedolan** [comment]
Precheck passed!
